langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java

```diff
@@ -193,72 +193,59 @@ private Future<Moderation> triggerModerationIfNeeded(Method method, List<ChatMes
         return (T) proxyInstance;
     }
 
-    private Optional<SystemMessage> prepareSystemMessage(Method method, Object[] args) {
-
-        Parameter[] parameters = method.getParameters();
-        Map<String, Object> variables = getPromptTemplateVariables(args, parameters);
+    private Optional<SystemMessage> prepareSystemMessage(Object memoryId, Method method, Object[] args) {
+        return prepareSystemMessageTemplate(memoryId, method)
+                .map(template -> PromptTemplate.from(template)
+                        .apply(getPromptTemplateVariables(args, method.getParameters()))
+                        .toSystemMessage());
+    }
 
+    private Optional<String> prepareSystemMessageTemplate(Object memoryId, Method method) {
         dev.langchain4j.service.SystemMessage annotation = method.getAnnotation(dev.langchain4j.service.SystemMessage.class);
         if (annotation != null) {
 
-            String systemMessageTemplate = getPromptText(
+            return Optional.of(getPromptText(
                     method,
                     "System",
                     annotation.fromResource(),
                     annotation.value(),
                     annotation.delimiter()
-            );
-
-            Prompt prompt = PromptTemplate.from(systemMessageTemplate).apply(variables);
-            return Optional.of(prompt.toSystemMessage());
+            ));
         }
 
-        return Optional.empty();
+        return context.systemMessagesProvider.apply(memoryId);
```

Comment:
With this new addition, we should extend the validation and fail if both `@XxxMessage` annotation and `xxxMessageProvider` are configured.

AI Service _**must**_ have either `@UserMessage` (either on top of method or on one of the parameters) or `userMessageProvider` configured, but not both at the same time.
AI Service _**can**_ have either `@SystemMessage` or `systemMessageProvider` configured, but not both at the same time.

Reply:
Done https://github.com/langchain4j/langchain4j/pull/862/commits/f33aa6140758bf7de30c47b7b622b17e1b28798f